docs/src/starknet/calldata-transformation.md

```diff
@@ -0,0 +1,172 @@
+> In general, paths to items have form: `<project-name>::<module-path>::<item-name>`.
+
+### Supported data types
+
+Cast supports expression of most important Cairo types:
+  * `bool`
+  * signed numbers (`i8`, `i16`, `i32`, `i64`, `i128`)
+  * unsigned numbers (`u8`, `u16`, `u32`, `u64`, `u128`, `u256`, `u512`)
+  * `felt252` (numeric literals and so-called *shortstrings*)
+  * `ByteArray`
+  * `ContractAddress`
+  * `ClassHash`
+  * `StorageAddress`
+  * `EthAddress`
+  * `bytes31`
+  * `Array` - using `array![]` macro
+
+Numeric types (primitives and `felt252`) can be paseed with type suffix specified - for example `--calldata 420_u64`.
+
+It's **not allowed** to write any function call expressions, either builtin, or user-defined.
+
+### More complex examples
+
+  1. `complex_fn` - different data types:
+
+  ```shell
+  $ sncast --account myuser \
+      call \
+      --url http://127.0.0.1:5050 \
+      --contract-address 0x016ad425af4585102e139d4fb2c76ce786d1aaa1cfcd88a51f3ed66601b23cdd \
+      --function tuple_fn \
+      --calldata \
+          array![array![1, 2], array![3, 4, 5], array![6]] \
+          12 \
+          -128_i8 \
+          "Some string (a ByteArray)" \
+          ('a shortstring', 32_u32) \
+          true \
+          0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff \
+      --block-id latest
+  ```
+
+  2. `complex_struct_fn` - struct nesting:
+
+  ```shell
+  $ sncast --account myuser \
+      call \
+      --url http://127.0.0.1:5050 \
+      --contract-address 0x016ad425af4585102e139d4fb2c76ce786d1aaa1cfcd88a51f3ed66601b23cdd \
+      --function tuple_fn \
+      --calldata \
+          data_transformer_contract::ComplexStruct { \
+              a: data_transformer_contract::SimpleStruct { a: 10 }, \
+              b: 12 \
+          } \
+      --block-id latest
+  ```
+
+### Caveats
+
+Reasoning about calldata serialization logic is not always possible for Cast. **In particular, Cast doesn't verify serialized calldata against the ABI**.
+
+When given a calldata, Cast first tries to transform it according to the contract ABI.
+When failed, in case of type or length mismatch, it will then try to interpret it as an already serialized data.
+
+Thus, calldata written in hex-encoded form is not always going to be interpreted as a list of felts and treated as serialized. Let's see an example:
+
+Our `DataTransformerContract` exposes a method with signature:
+```rust
+fn u256_fn(self: @ContractState, a: u256) { ... }
+```
+
+We can write:
+```shell
+$ sncast --account myuser \
+  call \
+  --url http://127.0.0.1:5050 \
+  --contract-address 0x016ad425af4585102e139d4fb2c76ce786d1aaa1cfcd88a51f3ed66601b23cdd \
+  --function u256_fn \
+  --calldata 0x10 \
+  --block-id latest
+```
+
+Despite the call data **not being** the proper serialization of an `u256` (which would be `0x10 0x0`), call succeeds.
```

Comment:
Maybe this should be a bit more strongly worded e.g. explicitly warn users to always use type suffixes etc. so their data doesn't become serialized incorrectly

Reply:
Also it would probably make sense to add a flag that disables data transformation just to be extra sure. This could be a separate issue and tackled by someone else though